camera/src/camera.mm

```diff
@@ -381,10 +380,65 @@ int CameraPlatform_StartCapture(dmBuffer::HBuffer* buffer, CameraType type, Capt
 
     g_Camera.m_VideoBuffer = *buffer;
 
-    return started ? 1 : 0;
+    if (started)
+    {
+        Camera_QueueMessage(STATUS_STARTED);
+    }
+    else
+    {
+        Camera_QueueMessage(STATUS_ERROR);
+    }
+}
+
+void CameraPlatform_StartCapture(dmBuffer::HBuffer* buffer, CameraType type, CaptureQuality quality, CameraInfo& outparams)
+{
+    // Only check for permission on iOS 7+ and macOS 10.14+
+    if ([AVCaptureDevice respondsToSelector:@selector(authorizationStatusForMediaType:)])
+    {
+        // Request permission to access the camera.
+        int status = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
+        if (status == AVAuthorizationStatusAuthorized)
+        {
+            // The user has previously granted access to the camera.
+            dmLogInfo("AVAuthorizationStatusAuthorized");
+            CameraPlatform_StartCaptureAuthorized(buffer, type, quality, outparams);
+        }
+        else if (status == AVAuthorizationStatusNotDetermined)
+        {
+            dmLogInfo("AVAuthorizationStatusNotDetermined");
+            // The app hasn't yet asked the user for camera access.
+            [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
+                if (granted) {
+                    dmLogInfo("AVAuthorizationStatusNotDetermined - granted!");
+                    CameraPlatform_StartCaptureAuthorized(buffer, type, quality, outparams);
```

Comment:
Now we have two places where we call "CameraPlatform_StartCaptureAuthorized".
It'd be nicer I think if we first check for status. and if it's not granted, then push a message and return. After that, we know it's granted, and we can call CameraPlatform_StartCaptureAuthorized at the end of the function.

Reply:
I didn't quite understand this. "push a message and return"?

Reply:
E.g.:

```
bool granted = false;
if (check1)
{
    if(check2)
    {
        granted = true;
    }
    else if (check3)
    {
            dmLogInfo("reason 1");
            Camera_QueueMessage(STATUS_NOT_PERMITTED);
    }
    else
    {
            dmLogInfo("Reason 2");
            Camera_QueueMessage(STATUS_NOT_PERMITTED);
    }
}

if (!granted)
    return;

CameraPlatform_StartCaptureAuthorized(buffer, type, quality, outparams);
```

In effect, only 1 call to `CameraPlatform_StartCaptureAuthorized` as opposed to 3.

Reply:
But one of the checks is asynchronous (pops dialog and calls back with result)

Reply:
Sure, I missed that since it's in the middle of an if statement with similar functionality :)
But for the first case, the argument could still be made (2 callsites is probably better than 3).
Oh well. Perhaps another time.